backends/oauth2.go

```diff
@@ -0,0 +1,335 @@
+	}
+
+	token, err := o.clientcredentialsConfig.Token(context.Background())
+
+	if err != nil {
+		log.Println(err)
+		return false, err
+	}
+
+	clientcredentialsClient := o.clientcredentialsConfig.Client(context.Background())
+
+	o.userCache[username] = userState{
+		username:  username,
+		superuser: false,
+		createdAt: time.Now(),
+		updatedAt: time.Unix(0, 0),
+		client:    clientcredentialsClient,
+		token:     token,
+	}
+
+	return true, err
+}
+
+func (o Oauth2) createUserWithToken(accessToken, clientid string) (bool, error) {
+	token := &go_oauth2.Token{
+		AccessToken: accessToken,
+		TokenType:   "Bearer",
+	}
+
+	client := o.oauth2Config.Client(context.Background(), token)
+
+	info, err := o.getUserInfo(client)
+
+	if err != nil {
+		log.Println(err)
+		return false, err
+	}
+
+	o.userCache[accessToken] = userState{
+		username:    accessToken,
+		superuser:   info.MQTT.Superuser,
+		createdAt:   time.Now(),
+		updatedAt:   time.Now(),
+		readTopics:  info.MQTT.Topics.Read,
+		writeTopics: info.MQTT.Topics.Write,
+		client:      client,
+		token:       token,
+	}
+
+	return true, err
+}
+
+func (o Oauth2) getUserInfo(client *http.Client) (*UserInfo, error) {
+	req, _ := http.NewRequest("GET", o.userInfoURL, nil)
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	info := UserInfo{}
+
+	err = json.NewDecoder(resp.Body).Decode(&info)
+	if err != nil {
+		return nil, err
+	}
+
+	return &info, nil
+}
+
+func (o Oauth2) checkAccessToTopic(topic string, acc int32, cache *userState, username string, clientid string) bool {
+	log.Debugf("Check for acl level %d", acc)
+
+	// check read access
+	if acc == 1 || acc == 4 {
+		res := o.isTopicInList(cache.readTopics, topic, username, clientid)
+		log.Debugf("ACL for read was %t", res)
+		return res
+	}
+
+	// check write
+	if acc == 2 {
+		res := o.isTopicInList(cache.writeTopics, topic, username, clientid)
+		log.Debugf("ACL for write was %t", res)
+		return res
+	}
+
+	// check for readwrite
+	if acc == 3 {
+		res := o.isTopicInList(cache.readTopics, topic, username, clientid) && o.isTopicInList(cache.writeTopics, topic, username, clientid)
+		log.Debugf("ACL for readwrite was %t", res)
+		return res
+	}
+	return false
+}
+
+func (o Oauth2) cacheIsValid(cache *userState) bool {
+	log.Debugf("Cache Expiary: %s", o.cacheDuration)
+	log.Debugf("Last Update: %s", cache.updatedAt)
+	log.Debugf("Difference to now: %s", time.Now().Sub(cache.updatedAt))
+
+	// function tests if the cache of the user is still valid
+	if o.cacheDuration == 0 {
+		return false
+	}
+
+	if (time.Now().Sub(cache.updatedAt)) < o.cacheDuration {
+		return true
+	}
+	return false
+}
+
+func (o Oauth2) isTopicInList(topicList []string, searchedTopic string, username string, clientid string) bool {
+	replacer := strings.NewReplacer("%u", username, "%c", clientid)
+
+	for _, topicFromList := range topicList {
+		if topics.Match(replacer.Replace(topicFromList), searchedTopic) {
```

Comment:
Huh, this made me realize that maybe replacing patterns could be done in more general way.